### ISSUE TITLE
Batch repo enhancements

### ISSUE THREAD

**Issue by o-alexandrov** [colinhacks/zod#225]
Let me please propose batch of changes to the DX, so it'd be easier for others to contribute.
Please let me know what you would approve, so I'd know what to help you with :)

### Todo (Not started yet)

- [ ] add [automated releases generation](https://medium.com/p/44c015f38fd6) and make commits messages more strict
  - as a user of this library, it was a bit time consuming to browse commits to realize what changed between `beta` versions

### Completed
- [x] use `typescript-eslint` instead of `tslint`
  - it's been almost 2 years since the standard in TS linting is `ESLint`, [more info](https://eslint.org/blog/2019/01/future-typescript-eslint)
- [x] add `.log` files to `.gitignore`
  - for example, there's already a mistakenly uploaded `yarn-error.log`
- [x] bump prettier & use community defaults
  - let's use the defaults, so it's more likely to please higher % of contributors
- [x] ~~add `.npmignore` to exclude unrelated files from the deployed package~~
  - improve the `files` + `exclude` instead
- [x] remove empty `dependencies` object from package.json
- [x] add `husky` to add `git` hooks
  - it's important to automatically prevent issues like in #218 from happening, when there are bunch of unrelated formatting changes
    - at least automatic code formatting should be in place
- [x] add official funding references of your GitHub sponsors link to the `package.json`
  - I suggest to start with the use described in [this PR from react-virtual](https://github.com/tannerlinsley/react-virtual/pull/78/files)

**Comment by colinhacks** [comment]
This all sounds great to me. I'm very open to a PR for these things. I'm also open to giving you commit access to the repo since you've been contributing to Zod for a long time, let me know 👍

**Comment by o-alexandrov** [comment]
Thank you. If you kindly grant commit access, I plan to carefully follow existing patterns in naming of commits, branches, and will always ask before doing any action for the first time.

I'm glad to contribute these changes in the next 48 hours.

**Comment by o-alexandrov** [comment]
Realized there are two other improvements, copied to the list above:
- add official funding references of your GitHub sponsors link to the `package.json`
  - I suggest to start with the use described in [this PR from react-virtual](https://github.com/tannerlinsley/react-virtual/pull/78/files)
- add [automated releases generation](https://medium.com/p/44c015f38fd6) and make commits messages more strict
  - as a user of this library, it was a bit time consuming to browse commits to realize what changed between `beta` versions

**Comment by jameschensmith** [comment]
@o-alexandrov, are you still working on this? I'd like to contribute on this if possible. @colinhacks, I could start getting separate PRs up for each of the points listed in the initial message. Let me know your thoughts when you get a chance, and I'll get the first PR pushed right away 😊

**Comment by o-alexandrov** [comment]
@jameschensmith please work on it, if you have time. I’m waiting on a reply for circular dependencies, before I work on this.

**Comment by jameschensmith** [comment]
Will do! 😁 I'll get started on it tomorrow morning 👍

**Comment by jameschensmith** [comment]
@o-alexandrov & @colinhacks,

Regarding the exclusion of unrelated files from the deployed package. There appears to be some funky things relating to yarn/npm & `.npmignore`. First, I came across [this article](https://medium.com/@jdxcode/for-the-love-of-god-dont-use-npmignore-f93c08909d8d), which recommends allow-listing (with `files` in `package.json`) over deny-listing first and foremost. Seeing as we are doing this this is good to see 🙂 

There is a problem, though, and it's quite messy. From the following two yarn issues reported ([1](https://github.com/yarnpkg/yarn/issues/685); [2](https://github.com/yarnpkg/yarn/issues/754)), I notice two problems when trying to use `.npmignore` or `.yarnignore` locally: (1) `lib/` is ignored by `.gitignore`, which because of its higher precedence is not ignoring any of the files within it; (2) if `files` is present in `package.json`, yarn apparently will only take allow-list rules (e.g. `!do-not-ignore-me.js`).

You can see the route that some have taken in the issues listed, but my recommendation is this: stick with `files`, and do the excludes from `tsconfig.base.json`.

Here's an example of what the exclude content could look like:

```json
{
  "exclude": [
    "./src/**/__tests__",
    "./src/playground.ts",
  ]
}
```

For examples & tests, we shouldn't need to compile them because we use tools such as TS Node (**bug:** There's a script with TS Node, but it's not in the dependencies) & TS Jest. Looking forward to your thoughts on which direction to take for this 😊 👍

**Comment by o-alexandrov** [comment]
@jameschensmith thank you very much, just to simplify the review for @colinhacks , I went through your PRs and all of what you did are improvements, imho.

Regarding excludes, also in agreement, in your PR, please also mark `/lib` instead of `lib` in `files` of `package.json`

**Comment by colinhacks** [comment]
@o-alexandrov I removed tslint entirely since I never actually used `yarn lint` and it wasn't integrated into the build process in any way. So perhaps the first item should be "Add eslint-typescript".

I checked off "funding". I also checked off "prettier" having just merged @jameschensmith's #260 

I also checked off "exclude". It's now using `includes` (not `files`) and `excludes` to prevent unused files from getting deployed. I just merged that change last night.
